packages/ingester/src/ingest.ts

```diff
@@ -0,0 +1,160 @@
+import {
+  EntityType,
+  Event,
+  EventType,
+  generateUniqueID,
+  Task,
+  TaskContent,
+  TaskIngestEvent,
+  TaskProvenance,
+  TaskUpdateDeletedEvent,
+} from "@withorbit/core";
+import { DatabaseEntityQuery, OrbitStore } from "@withorbit/store-shared";
+import {
+  IngestibleItem,
+  IngestibleItemIdentifier,
+  IngestibleSource,
+} from "./ingestible";
+
+type IngestOptions = {
+  ingestDateMilis: number;
+};
+const DEFAULT_OPTIONS = (): IngestOptions => ({
+  ingestDateMilis: Date.now(),
+});
+
+export const INGEST_ITEM_IDENTIFIER_KEY = "ingest_item_identifier" as const;
+
+// TODO: handle moves across sources
+// TODO: handle provenance updates
+export async function ingestSources(
+  sources: IngestibleSource[],
+  store: OrbitStore,
+  opts: IngestOptions = DEFAULT_OPTIONS(),
+) {
+  const events: Event[] = [];
+
+  // TODO: create a new query or extend the entity query such that
+  // we can filter on only sources that would have used this ingester
+  const query: DatabaseEntityQuery<Task> = {
+    entityType: EntityType.Task,
+  };
+  const entities = await store.database.listEntities(query);
+  const groupedEntities = groupEntitiesByProvenanceIdentifiers(entities);
```

Comment:
Reading this, I noticed that I had a lot of trouble following these different maps: what's being mapped to what? One thing I liked about Cocoa was its very consistent naming conventions—meant less thinking while coding, since names were largely formulaic, but also helped legibility. Maps/dictionaries/records used the convention `valuesByKey`, so the three maps here might be named: `existingEntitiesByProvenanceIdentifiers` (I added "existing" here because that distinction is important in this routine), `existingEntitiesByItemIdentifiers`, and `ingestibleItemsByItemIdentifiers`.

No obligation to change this, just thought you might find it interesting.

Reply:
Ah, the naming is complicated slightly by the fact that the first map is actually of _groups_; if this were Cocoa I'd probably name it `existingEntityGroupsByProvenanceIdentifiers` (of course, Cocoa likes long variable names, and that's perhaps less idiomatic in JS culture)